contracts/adapters/TargetWeightWrapExtension.sol

```diff
@@ -0,0 +1,382 @@
+     * @param _targetAsset Address of the target asset to wrap into.
+     * @param _reserveUnits Units of the reserve asset to wrap.
+     */
+    function wrap(
+        address _targetAsset,
+        uint256 _reserveUnits
+    )
+        external
+        nonReentrant
+        onlyAllowedRebalancer
+    {
+        require(isRebalancingActive, "Rebalancing is not active");
+        require(rebalanceInfo.targetAssets.contains(_targetAsset), "Invalid target asset");
+        require(isReserveOverweight(), "Reserve asset is not overweight");
+
+        bytes memory data = abi.encodeWithSelector(
+            wrapModule.wrap.selector,
+            setToken,
+            rebalanceInfo.reserveAsset,
+            _targetAsset,
+            _reserveUnits,
+            executionParams[_targetAsset].wrapAdapterName,
+            executionParams[_targetAsset].wrapData
+        );
+        invokeManager(address(wrapModule), data);
+
+        (uint256 targetAssetWeight, uint256 reserveWeight) = getTargetAssetAndReserveWeight(_targetAsset);
+        require(targetAssetWeight < executionParams[_targetAsset].maxTargetWeight, "Target asset overweight post-wrap");
+        require(reserveWeight > rebalanceInfo.minReserveWeight, "Reserve asset underweight post-wrap");
+    }
+
+    /**
+     * @notice Unwraps target asset units into the reserve asset.
+     * @dev Must be called when the reserve asset is underweight, rebalancing is enabled, and the caller is an allowed rebalancer.
+     * Ensures that after unwrapping, the target asset does not become underweight and the reserve asset does not become overweight.
+     * @param _targetAsset Address of target asset to unwrap from
+     * @param _targetUnits Units of target asset to unwrap
+     */
+    function unwrap(
+        address _targetAsset,
+        uint256 _targetUnits
+    )
+        external
+        nonReentrant
+        onlyAllowedRebalancer
+    {
+        require(isRebalancingActive, "Rebalancing is not active");
+        require(rebalanceInfo.targetAssets.contains(_targetAsset), "Invalid target asset");
+        require(isReserveUnderweight(), "Reserve asset is not underweight");
+
+        bytes memory data = abi.encodeWithSelector(
+            wrapModule.unwrap.selector,
+            setToken,
+            rebalanceInfo.reserveAsset,
+            _targetAsset,
+            _targetUnits,
+            executionParams[_targetAsset].wrapAdapterName,
+            executionParams[_targetAsset].unwrapData
+        );
+        invokeManager(address(wrapModule), data);
+
+        (uint256 targetAssetWeight, uint256 reserveWeight) = getTargetAssetAndReserveWeight(_targetAsset);
+        require(targetAssetWeight > executionParams[_targetAsset].minTargetWeight, "Target asset underweight post-unwrap");
+        require(reserveWeight < rebalanceInfo.maxReserveWeight, "Reserve asset overweight post-unwrap");
```

Comment:
Noting that this function doesn't limit rebalancers from unwrapping "too much" target asset, and hurting yield performance. This is less of an issue if we allow wrapping when the reserve is below its max but above minimum. See suggestion above in wrap().

Reply:
i agree, these functions make sense in conjunction with each other if the conditions are asymmetric and `wrap()` doesn't check reserve weight before, only after